src/dbtree.cpp

```diff
@@ -0,0 +1,343 @@
+                continue;
+            }
+
+            QString uuid = fileName.remove(36,3);
+            QString vPath = QString("/vm/") + uuid;
+            QStringList splitPath = vPath.split("/", QString::SplitBehavior::SkipEmptyParts);
+
+            qDebug() << "creating vms for uuid:" << uuid;
+
+            auto db = QSharedPointer<SimpleJsonDB>(new SimpleJsonDB(vPath, filePath, maxFlushDelayMillis));
+            vmsDbs.insert(uuid, db);
+
+            setValue(splitPath, db->getValue(), true);
+        }
+    }
+}
+
+DBTree::~DBTree()
+{
+}
+
+QSharedPointer<SimpleJsonDB> DBTree::lookupDb(const QStringList &splitPath)
+{
+    if (splitPath.length() < 2) {
+        return mainDb;
+    }
+
+    QString topLevel = splitPath[0];
+    QString secondLevel = splitPath[1];
+
+    if (topLevel == "dom-store") {
+        QHash<QString, QSharedPointer<SimpleJsonDB>>::iterator i = domstoreDbs.find(secondLevel);
+        if (i != domstoreDbs.end() && i.key() == secondLevel) {
+            return i.value();
+        }
+
+        // does not exist - create one
+        QString filePath = dbPath + QDir::separator() + "dom-store" + QDir::separator() + secondLevel + ".db";
+        QString vPath = QString("/dom-store/") + secondLevel;
+        QStringList baseSplitPath;
+        baseSplitPath << topLevel << secondLevel;
+
+        qDebug() << "creating dom-store db for uuid:" << secondLevel << "at path:" << filePath;
+        auto db = QSharedPointer<SimpleJsonDB>(new SimpleJsonDB(vPath, filePath, maxFlushDelay));
+        domstoreDbs.insert(secondLevel, db);
+        setValue(baseSplitPath, db->getValue(), true);
+        return db;
+    } else if (topLevel == "vm") {
+        QHash<QString, QSharedPointer<SimpleJsonDB>>::iterator i = vmsDbs.find(secondLevel);
+        if (i != vmsDbs.end() && i.key() == secondLevel) {
+            return i.value();
+        }
+
+        // does not exist - create one
+        QString filePath = dbPath + QDir::separator() + "vms" + QDir::separator() + secondLevel + ".db";
+        QString vPath = QString("/vm/") + secondLevel;
+        QStringList baseSplitPath;
+        baseSplitPath << topLevel << secondLevel;
+
+        qDebug() << "creating vm db for uuid:" << secondLevel << "at path:" << filePath;
+        auto db = QSharedPointer<SimpleJsonDB>(new SimpleJsonDB(vPath, filePath, maxFlushDelay));
+        vmsDbs.insert(secondLevel, db);
+        setValue(baseSplitPath, db->getValue(), true);
+        return db;
+    }
+
+    // if not in domstore or vms db, must be in main db
+    return mainDb;
+}
+
+QMPointer<QMJsonValue> DBTree::getValue(const QStringList &splitPath)
+{
+    QMPointer<QMJsonValue> obj = dbRoot;
+
+    qDebug() << "getValue(): splitPath:" << splitPath;
+
+    // if it is top of tree, return the whole tree
+    if (splitPath.length() == 0) {
+        return obj;
+    }
+
+    // traverse tree parts
+    QStringList currentSplitPath;
+    foreach (const QString &part, splitPath) {
```

Comment:
I would use a range for instead of foreach since your not modifying the list. It's a little faster since it doesn't have to do the copy